packages/ejson/lib/src/decoding.dart

```diff
@@ -0,0 +1,272 @@
+  final type = T;
+  final nullable = type.isNullable;
+  final decoder = nullable ? _decodeNullable : decoders[type.base];
+  if (decoder == null) {
+    throw MissingDecoder(ejson, type);
+  }
+  final args = nullable ? [type.nonNull] : type.args;
+  if (args.isEmpty) {
+    return decoder(ejson) as T; // minor optimization
+  }
+  return decoder.callWith(typeArguments: args, parameters: [ejson]) as T;
+}
+
+// Important to return `T` as opposed to [Never] for type inference to work
+/// @nodoc
+T raiseInvalidEJson<T>(Object? value) => throw InvalidEJson(value, T);
+
+dynamic _decodeAny(EJsonValue ejson) {
+  return switch (ejson) {
+    null => null,
+    bool b => b,
+    double d => d, // relaxed mode
+    int i => i, // relaxed mode
+    String s => s,
+    {'\$date': _} => _decodeDate(ejson),
+    {'\$maxKey': _} => _decodeKey(ejson),
+    {'\$minKey': _} => _decodeKey(ejson),
+    {'\$numberDouble': _} => _decodeDouble(ejson),
+    {'\$numberInt': _} => _decodeInt(ejson),
+    {'\$numberLong': _} => _decodeInt(ejson),
+    {'\$regex': _} => _decodeString(ejson),
+    {'\$symbol': _} => _decodeSymbol(ejson),
+    {'\$undefined': _} => _decodeUndefined<dynamic>(ejson),
+    {'\$oid': _} => _decodeObjectId(ejson),
+    {'\$binary': {'base64': _, 'subType': '04'}} => _decodeUuid(ejson),
+    {'\$binary': _} => _decodeBinary(ejson),
+    List<dynamic> _ => _decodeArray<dynamic>(ejson),
+    Map<dynamic, dynamic> _ => _tryDecodeCustom(ejson) ?? _decodeDocument<String, dynamic>(ejson), // other maps goes last!!
+    _ => raiseInvalidEJson<dynamic>(ejson),
+  };
+}
+
+dynamic _tryDecodeCustom(EJsonValue ejson) {
+  for (final decoder in customDecoders.values) {
+    try {
+      return decoder(ejson);
+    } catch (_) {
+      // ignore
+    }
+  }
+  return null;
+}
+
+List<T> _decodeArray<T>(EJsonValue ejson) {
+  return switch (ejson) {
+    Iterable<dynamic> i => i.map((ejson) => fromEJson<T>(ejson)).toList(),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+bool _decodeBool(EJsonValue ejson) {
+  return switch (ejson) {
+    bool b => b,
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+DateTime _decodeDate(EJsonValue ejson) {
+  return switch (ejson) {
+    {'\$date': String s} => DateTime.parse(s), // relaxed mode
+    {'\$date': {'\$numberLong': String i}} => DateTime.fromMillisecondsSinceEpoch(int.tryParse(i) ?? raiseInvalidEJson(ejson)),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+Defined<T> _decodeDefined<T>(EJsonValue ejson) {
+  if (ejson case {'\$undefined': 1}) return raiseInvalidEJson(ejson);
+  return Defined<T>(fromEJson<T>(ejson));
+}
+
+Map<K, V> _decodeDocument<K, V>(EJsonValue ejson) {
+  return switch (ejson) {
+    Map<dynamic, dynamic> m => m.map((key, value) => MapEntry(key as K, fromEJson<V>(value))),
+    _ => raiseInvalidEJson(ejson),
+  };
+}
+
+double _decodeDouble(EJsonValue ejson) {
+  return switch (ejson) {
+    double d => d, // relaxed mode
+    //{'\$numberDouble': double d} => d,
```

Comment:
Should we delete this code?

Reply:
Yes, it is neither `canonical` or `relaxed` formatting, but something in-between.

Reply:
Fixed